Search tags on expression functions

signedav opened the pull request:
Tags on static expression functions for better find of the used function.

Instead of only be able to find `array_to_string` with content of this descriptive name, we can find it now with e.g. "split" or "convert" etc.

![searchtags](https://user-images.githubusercontent.com/28384354/70033165-2dcb9c00-15af-11ea-98a9-3a4305dd9180.gif)

It's not implemented with aliases to avoid having so many different ways to call the same function. With the search tags we are much more flexible.

signedav committed:
search tags in expression function to look up the functionality in the expression builder

signedav committed:
search tags on some static expression functions

m-kuhn commented on src/core/expression/qgsexpressionfunction.cpp:
```diff
@@ -5961,8 +5962,8 @@ const QList<QgsExpressionFunction *> &QgsExpression::Functions()
         << new QgsStaticExpressionFunction( QStringLiteral( "array_length" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "array" ) ), fcnArrayLength, QStringLiteral( "Arrays" ) )
-        << new QgsStaticExpressionFunction( QStringLiteral( "array_contains" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "array" ) ) << QgsExpressionFunction::Parameter( QStringLiteral( "value" ) ), fcnArrayContains, QStringLiteral( "Arrays" ) )
-        << new QgsStaticExpressionFunction( QStringLiteral( "array_all" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "array_a" ) ) << QgsExpressionFunction::Parameter( QStringLiteral( "array_b" ) ), fcnArrayAll, QStringLiteral( "Arrays" ) )
+        << new QgsStaticExpressionFunction( QStringLiteral( "array_contains" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "array" ) ) << QgsExpressionFunction::Parameter( QStringLiteral( "value" ) ), fcnArrayContains, QStringLiteral( "Arrays" ), QString(), false, QSet<QString>(), false, QStringList(), false, QStringList() << QStringLiteral( "exist ...
```
Can the tags be made translatable?

DelazJ replied:
I don't know whether it's the same context but when translating, Processing tags are shown as a comma separated list, eg `join,connect,attributes,values,fields,tables` that one could translate into `join,connect,attributes,values,fields,tables,joindre,connecter,valeurs,champs` convenient when you remember a tag in English while on a French QGIS. This is because the string is identifiable as a tag list and not as separate words one would need to translate. Is that kept or possible to keep?

signedav replied:
I plan to change it a little bit (see my last commit) and the tags will be held in the `Help` object, so the function is created regarding the files defined in `resources/function_help/json/`

When there is defined: 
```
"tags": [ "split", "convert", "seperate" ]
```
A function is created in `qgsexpression_texts.cpp` with a list like:
```
            QStringList()
              << QStringLiteral( "split" ) << tr( "split" )
              << QStringLiteral( "convert" ) << tr( "convert" )
              << QStringLiteral( "seperate" ) << tr( "seperate" )
```
So you translate it like usual and the English tag is kept.

m-kuhn commented on src/gui/qgsexpressionbuilderwidget.cpp:
```diff
@@ -1231,7 +1232,21 @@ bool QgsExpressionItemSearchProxy::filterAcceptsRow( int source_row, const QMode
+  }
+  else
+  {
+    QStringList tags = sourceModel()->data( index, QgsExpressionItem::SEARCH_TAGS_ROLE ).toStringList();
```
```suggestion
    const QStringList tags = sourceModel()->data( index, QgsExpressionItem::SEARCH_TAGS_ROLE ).toStringList();
```

m-kuhn reviewed (commented):
Nice.
One suggestion, could this simply be named "tags" instead of "searchTags"?

signedav committed:
Update src/gui/qgsexpressionbuilderwidget.cpp

Co-Authored-By: Matthias Kuhn <matthias@opengis.ch>

signedav commented:
> One suggestion, could this simply be named "tags" instead of "searchTags"

Had it like that first, but prefered `searchTags` then, because it defines that it's (compared to other attributes like name and aliases) only used for the search part of it... But if you insist I can change it, I don't mind that much :-)

m-kuhn commented:
There is already `QgsProcessingAlgorithm::tags`, so it would keep the API coherent.
Plus it would avoid using it like this by accident: `searchTags( 'the tag I want to search for' )` because `tags` is clearly a getter while `searchTags` can be understood an imperative command.

signedav commented:
Okay. With this context I agree @m-kuhn

NathanW2 commented:
Good idea!

On Tue., 3 Dec. 2019, 6:52 pm signedav, <notifications@github.com> wrote:

> Okay. With this context I agree @m-kuhn <https://github.com/m-kuhn>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/33197?email_source=notifications&email_token=AAC5FXHBX2UF5JWOMRMMV4LQWYM3VA5CNFSM4JUT7N72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFYSQLA#issuecomment-561063980>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXGKQNBLCDHIJ25HZWTQWYM3VANCNFSM4JUT7N7Q>
> .
>

signedav committed:
tags helping to find the correct function are integrated into the help of the expression
defined in the json files and used by the expression builder to find the function (not using as alias)

signedav committed:
write tags to the expression functions help object

m-kuhn commented on python/core/auto_generated/expression/qgsexpression.sip.in:
```diff
@@ -535,6 +535,15 @@ Returns the help text for a specified function.
 .. seealso:: :py:func:`formatVariableHelp`
+%End
+
+    static QStringList tags( QString name );
```
```suggestion
    static QStringList tags( const QString &name );
```

signedav replied:
oh that's the sip...

signedav committed:
Update python/core/auto_generated/expression/qgsexpression.sip.in

Co-Authored-By: Matthias Kuhn <matthias@opengis.ch>

and

set parameter to const

signedav committed:
fix spelling

signedav committed:
tags instead of searchTags

signedav committed:
remove unneeded set of default value null

skampus commented:
to facilitate the translation, I ask you if it is possible to use the format previously adopted for the tags, for example (clip,extract,intersect,intersection,mask,extent).
This allows us translators to identify the tags at a glance and decide whether to translate them or not.
In Italian we are oriented not to translate them, since the same words are contained in the contextual guide and therefore the functions are in any case found.
Do you think it is possible to adopt this method?

is it better to ask in dev list?

m-kuhn commented:
Please go ahead and open a pull request with an example, best to discuss with that. You can do that also with the github web editor directly.
Thanks

agiudiceandrea commented:
Hi @signedav, it seems to me the tags are not used for Operators in searches: e.g. searching for "multiplication" doesn't find the "*" operator, or searching for "division" doesn't find the operator "/" and the operator "%". Is this intended or not yet implemented?

agiudiceandrea commented:
Thanks!. I've opened an issue report https://github.com/qgis/QGIS/issues/52157.
